### DISCUSSION TITLE
arrayBuffer->Detach() in C++ has more overhead than it was expected to.

### DISCUSSION THREAD

**Discussion by Guthib-of-Dan** [uNetworking/uWebSockets.js#1266] [Ideas]
I have recently finished creating 2 TC39 proposals on the way we handle network data in JavaScript - "JSON.parseBinary" and "ArrayBuffer.prototype.detach" (the one I am going to talk about now).
> Let me promote these. I want github stars: https://github.com/Guthib-of-Dan/proposal-arraybuffer-detach; https://github.com/Guthib-of-Dan/proposal-json-parse-binary/

Both proposals contain benchmarks to make a solid background for the idea and "proposal-arraybuffer-detach" has an example with C++ V8 API and NAPI, where I call some js callback in a loop from C++, give it an ArrayBuffer over existing memory and detach that memory after the callback (or IN the callback). Just like uWebSockets.js does.  
The nuance is that results do have a difference, which is unnoticeable at first.  
```
V8 API
──────────────────────────────────────────────────────────────────
  warmup                      ██████████░░░░░░░░░░░░░░░░░░    9.297 s
  C++ detach via JS call      █████████░░░░░░░░░░░░░░░░░░░    8.067 s  8.0% faster than C++ auto-detach  ⚑
  C++ detach after callback   █████████░░░░░░░░░░░░░░░░░░░    8.766 s  C++ post-callback detach
  internalDetach()            ████████░░░░░░░░░░░░░░░░░░░░    7.945 s
  C++ detach after cb (2nd)   █████████░░░░░░░░░░░░░░░░░░░    8.935 s  C++ post-callback detach
  internalDetach() (2nd)      ████████░░░░░░░░░░░░░░░░░░░░    8.023 s
  no detach                   ████████████████████████████   26.546 s  3.2× slower than avg detach
```
What an AI suggest is that inside the callback "ArrayBuffer" as a param is kept hot in memory, but after callback - not. Also this is not just about speed, this is about immediately handling the memory you know you needn't after some parsing (gc-pressure test in that repo suggest that leaving GC handle all that memory comes at a huge cost).

### What should you do with this?
So we can create new handlers or extend existing functionality to support manual memory management. Everything, obviously, has a sense of integrating only if at least "proposal-arraybuffer-detach" gets its deserved appraisal.

### Btw, don't you mind me mentioning uWebSockets.js in my proposal? 
It is already there in README, but this discussion is the first place where it goes to. Do you have any objections?

**Comment by uNetworkingAB** [comment]
> this is about immediately handling the memory you know you needn't after some parsing (gc-pressure test in that repo suggest that leaving GC handle all that memory comes at a huge cost).

Lol. Well, yes.

Rust, C, C++ are fast **because** they inherently do not have a GC. Go, Java, JavaScript, Haskell are slow **because** they are inherently designed around a GC.

Proposing manual memory management for JavaScript is like proposing Atheism to be more like Islam. It's the wrong venue.

**Reply by Guthib-of-Dan** [reply]
I mean it speeds things up, doesn’t it? At least in the benchmark.

**Reply by uNetworkingAB** [reply]
ArrayBuffer::Detach is not used by uWS.js to speed things up. It's used for safety.

Detach used to be called Neuter and it essentially just set the length of the ArrayBuffer to 0:

<img width="1370" height="192" alt="Screenshot 2026-03-30 at 23 20 36" src="https://github.com/user-attachments/assets/b80a2fdd-2850-465b-bc4e-49211271daf9" />

uWS.js does not put the content of the ArrayBuffer in the GC, we only pass non-owning ArrayBuffers that "wrap" stack memory. The only reason we Neuter / Detach is to reset the ArrayBuffer so that anyone who keeps it past the return of the callback simply see 0 length and not garbled up data. So for us it's a security feature.

**Reply by Guthib-of-Dan** [reply]
So that’s a “no” for extending existing functionality. I guess that same applies to creating new? Like another callback for “websockets” as an example

app.ws(“/“, {
  open(){},
// either “message” or “messageManual”
  messageManual(ws, msg){
    var json = JSON.parseBinary(msg);
    msg.detach();
    // do other stuff
  }
})

if still no, then should I remove mentioning uWS.js from the proposal?

**Reply by uNetworkingAB** [reply]
> I mean it speeds things up, doesn’t it? At least in the benchmark.

It's like saying Ice cream tastes too much like ice cream I want it to taste like Broccoli because Broccoli is better, in a convention about Ice cream instead of just buying a ticket to the broccoli convention.

**Reply by Guthib-of-Dan** [reply]
Alright. So the last question
> should I remove mentioning uWS.js from the proposal?

**Reply by uNetworkingAB** [reply]
> So that’s a “no” for extending existing functionality.

I never said that, and I have zero authority over what you want to propose. Propose whatever you want. I'm just saying, generally speaking, that most people are retarded and won't agree with your opinion.

My opinion is that JS should a) remove the GC entirely b) remove dynamic typing and have only static types and c) be replaced in its entirety with something entirely different. But 99% of programmers disagree because of before mentioned retardation.

**Reply by Guthib-of-Dan** [reply]
And still, I don’t try to bring something groundbreaking to js. I just want one function that does a little less work than its equivalent .transfer(0). No one forbids using it now.

**Reply by Guthib-of-Dan** [reply]
And yes, I share your thoughts considering someone’s high-level knowledge and “React-only” experience. But not trying is worse than trying and getting a rejection, isn’t it?

**Comment by marco76tv** [comment]
## Decision log — 2026-05-28

**Decisione:** `/it` (slug `home`) = pagina «Elenco segnalazioni» Design Comuni, composta da blocchi CMS universali; breadcrumb fuori JSON; griglia 2 col con filtri a sinistra sempre presenti su desktop.

**Motivazione:** parity reference + DRY (discussion #36) + regola no-page-specific blocks (#77).

**Dettaglio:** vedi commento in [discussion #36](https://github.com/laraxot/base_fixcity_fila5/discussions/36#discussioncomment-17101339) e STORY-062 nel mono.

**Owner:** theme_sixteen (blade/CMS), module_geo (tickets.json/map-lit), module_fixcity (ViewModel filtri), module_cms (x-page breadcrumb).
